File: ngo/ngo/report/duplicate_finder/duplicate_finder.py
# Copyright (c) 2024, smb and contributors
# For license information, please see license.txt

import frappe
import pandas as pd


def execute(filters=None):
	columns = [
		{
			'fieldname': 'donor_id_number',
			'label': frappe._('Donor Id'),
			'fieldtype': 'Link',
			'options': 'Donor'
		},
		{
			'fieldname': 'donor_id_number_count',
			'label': frappe._('Donor Id Count'),
			'fieldtype': 'Int'
		},
		{
			'fieldname': 'account_no',
			'label': frappe._('Bank Account'),
			'fieldtype': 'Link',
			'options': 'Bank Account'
		},
		{
			'fieldname': 'account_no_count',
			'label': frappe._('Bank Account Count'),
			'fieldtype': 'Int'
		},
		{
			'fieldname': 'parent',
			'label': frappe._('Slip Form'),
			'fieldtype': 'Link',
			'options': 'Slip Form'
		},
		{
			'fieldname': 'srno',
			'label': frappe._('Cheque Serial Number'),
			'fieldtype': 'Int'
		}
	]
	data = []

	print(filters, 'filters')

	slip_forms = []
	query_filters = {}
	if 'event' in filters:
		slip_forms = [i.name for i in frappe.db.get_list('Slip Form', filters={'slip_event_code': filters['event']})]
	
	if len(filters['slip_forms']) > 0:
		slip_forms = filters['slip_forms']

	if len(slip_forms) > 0:
		query_filters['parent'] = ['IN', slip_forms]

	data_ = frappe.db.get_list('Slip Cheque Form', fields=['donor_id_number', 'account_no', 'parent', 'srno'], page_length=100000, filters=query_filters, debug=True)
	print(data_, 'data_', len(data_))
	
	df = pd.DataFrame.from_records(data_)
	df['donor_id_number_count'] = df.groupby(['donor_id_number'])['donor_id_number'].transform('count')
	df['account_no_count'] = df.groupby(['account_no'])['account_no'].transform('count')
	data = df.loc[(df['donor_id_number_count'] > 1) | (df['account_no_count'] > 1)].to_dict('records')

	
	return columns, data

File: ngo/ngo/report/duplicate_finder/duplicate_finder.js
// Copyright (c) 2024, smb and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Duplicate Finder"] = {
	"filters": [
		{
			fieldname: 'event',
			label: __('Event'),
			fieldtype: 'Link',
			options: 'Event Master',
			on_change: function() {
				frappe.query_report.set_filter_value('slip_forms', [])
			}
		},
		{
			"fieldname": "slip_forms",
			"label": __("Slip Forms"),
			"fieldtype": "MultiSelectList",
			"get_data": function(txt) {
				var filters = {}
				if(frappe.query_report.get_filter_value('event').length > 0) {
					filters['slip_event_code'] = frappe.query_report.get_filter_value('event')
				}
				return frappe.db.get_link_options("Slip Form", txt, filters);
			}
		}
	]
};
